File: algorithms/visualstudio/algorithms.fsharp.tests/MergeSortCountInversionsTests.fs
namespace algorithms.fSharp.divideAndConquer.tests

open Microsoft.VisualStudio.TestTools.UnitTesting
open algorithms.fSharp.divideAndConquer.MergeSortCountInversions
open algorithms.fSharp.utilities.ParseTextFile


[<TestClass>]
type MergeSortCountInversions() = 

    [<TestMethod>]
    member x.``MergeSortCountInversions F# ListAndInversions type [1] should return item with 0 as default numberOfInversions`` () = 
        Assert.AreEqual(0, ListAndInversions([1]).numberOfInversions)

    [<TestMethod>]
    member x.``MergeSortCountInversions F# ListAndInversions type [1] 2 should return item with 2 numberOfInversions`` () = 
        Assert.AreEqual(2, ListAndInversions([1], 2).numberOfInversions)

    [<TestMethod>]
    member x.``MergeSortCountInversions F# mergeAndCountInversions [1] [] should return sortedArray`` () = 
        x.testIfMergeAndCountInversionsReturnsASortedArray([1],[])

    [<TestMethod>]
    member x.``MergeSortCountInversions F# mergeAndCountInversions [1] [] should return 0 inversions`` () = 
        x.testIfMergeAndCountInversionsReturnsCorrectInversions([1],[],0)

    [<TestMethod>]
    member x.``MergeSortCountInversions F# mergeAndCountInversions [1;3;5] [2;4;6] should return sortedArray`` () = 
        x.testIfMergeAndCountInversionsReturnsASortedArray([1;3;5],[2;4;6])

    [<TestMethod>]
    member x.``MergeSortCountInversions F# mergeAndCountInversions [1;3;5] [2;4;6] should return 3 inversions`` () = 
        x.testIfMergeAndCountInversionsReturnsCorrectInversions([1;3;5],[2;4;6],3)

    [<TestMethod>]
    member x.``MergeSortCountInversions F# mergeAndCountInversions [7;8;9;13;15] [1;2;5;7;8] should return sortedArray`` () = 
         x.testIfMergeAndCountInversionsReturnsASortedArray([7;8;9;13;15],[1;2;5;7;8])

    [<TestMethod>]
    member x.``MergeSortCountInversions F# mergeAndCountInversions [7;8;9;13;15] [1;2;5;7;8] should return 22 inversions`` () = 
        x.testIfMergeAndCountInversionsReturnsCorrectInversions([7;8;9;13;15],[1;2;5;7;8],22)

    member x.testIfMergeAndCountInversionsReturnsCorrectInversions(list1, list2, expectedInversions) = 
        let listAndInversions1 = ListAndInversions(list1)
        let listAndInversions2 = ListAndInversions(list2)

        let result = mergeAndCountInversions(listAndInversions1, listAndInversions2)

        Assert.AreEqual(expectedInversions, result.numberOfInversions)

    member x.testIfMergeAndCountInversionsReturnsASortedArray(list1, list2) = 
        let listAndInversions1 = ListAndInversions(list1)
        let listAndInversions2 = ListAndInversions(list2)
        
        let expectation = List.sort <| List.append list1 list2 

        let result = mergeAndCountInversions (listAndInversions1, listAndInversions2)        
        
        Assert.AreEqual(expectation, result.list)

    [<TestMethod>]
    member x.``MergeSortCountInversions F# countInversions [6..1] should return 15 inversions`` () = 
        let result = [1..6] |> List.rev |> countInversions
        Assert.AreEqual(15, result)

    [<TestMethod>]
    member x.``MergeSortCountInversions F# countInversions [6;5] should return 1 inversions`` () = 
        let result = [6;5] |> countInversions
        Assert.AreEqual(1, result)

    [<TestMethod>]
    member x.``MergeSortCountInversions F# countInversions [1000..1] should return 499500 inversions`` () = 
        let result = [1..1000] |> List.rev |> countInversions
        Assert.AreEqual(499500, result)

    [<TestMethod>]
    member x.``MergeSortCountInversions F# countInversions [1..10] should return 0 inversions`` () = 
        Assert.AreEqual(0, countInversions [1..10])

    [<TestMethod>]
    member x.``MergeSortCountInversions F# countInversions [1;3;5;2;4;6] should return 3 inversions.`` () = 
        Assert.AreEqual(3, countInversions [1;3;5;2;4;6])

File: algorithms/visualstudio/algorithms.csharp.tests/DivideAndConquer/MergeSortCountInversionsTests.cs
using System.Collections.Generic;
using System.Linq;
using algorithms.csharp.DivideAndConquer;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using algorithms.fSharp.utilities;

namespace algorithms.csharp.tests.DivideAndConquer
{
    [TestClass]
    public class MergeSortCountInversionsTests
    {
        private MergeSortCountInversions _countInversions;

        [TestInitialize]
        public void Initialize()
        {
            _countInversions = new MergeSortCountInversions();
        }

        #region MergeAndCountInversions

        [TestMethod]
        public void MergeSortCountInversions_CSharp_MergeAndCountInversions_1_Empty_Should_Return_SortedArray()
        {
            var list1 = Enumerable.Range(1, 1).ToList();
            var list2 = new List<int>();

            MergeListSetupAndTest(list1, list2);
        }

        [TestMethod]
        public void MergeSortCountInversions_CSharp_MergeAndCountInversions_1_Empty_Should_Return_0_Inversions()
        {
            var list1 = Enumerable.Range(1, 1).ToList();
            var list2 = new List<int>();

            CountOfInversionsFromMergeSetupAndTest(list1, list2, 0);
        }

        [TestMethod]
        public void MergeSortCountInversions_CSharp_MergeAndCountInversions_135_246_Should_Return_SortedArray()
        {
            var list1 = new List<int> {1, 3, 5};
            var list2 = new List<int> {2, 4, 6};

            MergeListSetupAndTest(list1, list2);
        }

        [TestMethod]
        public void MergeSortCountInversions_CSharp_MergeAndCountInversions_135_246_Should_Return_3_Inversions()
        {
            var list1 = new List<int> {1, 3, 5};
            var list2 = new List<int> {2, 4, 6};

            CountOfInversionsFromMergeSetupAndTest(list1, list2, 3);
        }

        [TestMethod]
        public void MergeSortCountInversions_CSharp_MergeAndCountInversions_7891315_12578_Should_Return_SortedArray()
        {
            var list1 = new List<int> {7, 8, 9, 13, 15};
            var list2 = new List<int> {1, 2, 5, 7, 8};

            MergeListSetupAndTest(list1, list2);
        }

        [TestMethod]
        public void MergeSortCountInversions_CSharp_MergeAndCountInversions_7891315_12578_Should_Return_22_Inversions()
        {
            var list1 = new List<int> {7, 8, 9, 13, 15};
            var list2 = new List<int> {1, 2, 5, 7, 8};

            CountOfInversionsFromMergeSetupAndTest(list1, list2, 22);
        }

        private void CountOfInversionsFromMergeSetupAndTest(List<int> list1, List<int> list2, int expectedInversions)
        {
            var result = _countInversions.MergeAndCountInversions(new ListAndInversions(list1),
                new ListAndInversions(list2));

            Assert.AreEqual(expectedInversions, result.CountOfInversions);
        }

        private void MergeListSetupAndTest(List<int> list1, List<int> list2)
        {
            var expected = new List<int>(list1);
            expected.AddRange(list2);
            expected.Sort();

            var result = _countInversions.MergeAndCountInversions(new ListAndInversions(list1),
                new ListAndInversions(list2));

            Assert.IsTrue(result.List.SequenceEqual(expected));
        }

        #endregion

        [TestMethod]
        public void MergeSortCountInversions_CSharp_CountInversions_6To1_Should_Return_15()
        {
            var list = Enumerable.Range(1, 6).Reverse().ToList();

            Assert.AreEqual(15, _countInversions.CountInversions(list));
        }

        [TestMethod]
        public void MergeSortCountInversions_CSharp_CountInversions_6To5_Should_Return_1()
        {
            var list = Enumerable.Range(5, 2).Reverse().ToList();

            Assert.AreEqual(1, _countInversions.CountInversions(list));
        }

        [TestMethod]
        public void MergeSortCountInversions_CSharp_CountInversions_1000To1_Should_Return_1()
        {
            var list = Enumerable.Range(1, 1000).Reverse().ToList();

            Assert.AreEqual(499500, _countInversions.CountInversions(list));
        }

        [TestMethod]
        public void MergeSortCountInversions_CSharp_CountInversions_1To10_Should_Return_0()
        {
            var list = Enumerable.Range(1, 10).ToList();

            Assert.AreEqual(0, _countInversions.CountInversions(list));
        }

        [TestMethod]
        public void MergeSortCountInversions_CSharp_CountInversions_135246_Should_Return_3()
        {
            var list = new List<int>{1,3,5,2,4,6};

            Assert.AreEqual(3, _countInversions.CountInversions(list));
        }

        [TestMethod]
        public void MergeSortCountInversions_CSharp_CountInversions_Homework_Should_Return_Answer()
        {
            var list = ParseTextFile.parseNewLineDelimitedArray("C:\\Users\\Thomas\\Documents\\GitHub\\algorithms.practice\\algorithms.practice\\problemSets\\IntegerArray.txt").ToList();

            Assert.AreEqual(2407905288, _countInversions.CountInversions(list));
        }
    }
}
